sklearn/feature_extraction/text.py

```diff
@@ -1068,17 +1068,21 @@ def fit(self, X, y=None):
         """
         if not sp.issparse(X):
             X = sp.csc_matrix(X)
+
         if self.use_idf:
             n_samples, n_features = X.shape
-            df = _document_frequency(X)
+            # _document_frequency will return int32
+            # array by default, hence the cast
+            df = _document_frequency(X).astype(X.dtype
+                                               if X.dtype in FLOAT_DTYPES
+                                               else np.float64)
 
             # perform idf smoothing if required
             df += int(self.smooth_idf)
             n_samples += int(self.smooth_idf)
-
             # log+1 instead of log makes sure terms with zero idf don't get
             # suppressed entirely.
-            idf = np.log(float(n_samples) / df) + 1.0
+            idf = np.log(n_samples / df) + 1.0
```

Comment:
I would have thought we should always perform this log in float64, and only afterwards convert back to float32 if appropriate. Confused. I don't understand why you had that test failing before and not now.

Reply:
Because `df` is converted to float if it's in int

Reply:
But shouldn't it be converted to float64 in all cases, then idf converted back to float32 if necessary?

Reply:
Yeah it seems a little magical that the dtype after transforming matches even if I am not converting it back to the original type. I'll do a little debugging to answer this question.

Reply:
But you're also unnecessarily reducing the precision of the calculation here in the float32 case.

Reply:
OK so the required behavior is always convert `df` to float64 for precision and convert `idf` back to X.dtype?

Reply:
It was my proposal in https://github.com/glemaitre/scikit-learn/commit/4e5e1f06ed1b90c0ffb00584db81a4e8c77e1dff